File: app/src/main/java/com/metinkale/prayerapp/vakit/LocationService.java
/*
 * Copyright (c) 2013-2017 Metin Kale
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.metinkale.prayerapp.vakit;

import android.Manifest;
import android.app.Service;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.location.Location;
import android.location.LocationListener;
import android.location.LocationManager;
import android.os.Bundle;
import android.os.IBinder;
import android.os.Looper;
import android.support.annotation.IntDef;
import android.support.v4.app.ActivityCompat;
import android.widget.Toast;

import com.metinkale.prayerapp.utils.Geocoder;
import com.metinkale.prayerapp.vakit.times.Cities;
import com.metinkale.prayerapp.vakit.times.Entry;
import com.metinkale.prayerapp.vakit.times.Source;
import com.metinkale.prayerapp.vakit.times.Times;
import com.metinkale.prayerapp.vakit.times.sources.CalcTimes;
import com.metinkale.prayerapp.vakit.times.sources.WebTimes;

import java.util.List;

@SuppressWarnings("MissingPermission")
public class LocationService extends Service implements LocationListener {
    private static final String UPDATE_LOCATION = "UPDATE_LOCATION";
    private LocationManager mLocationManager;
    private boolean mHasWebTime = false;
    private boolean mHasCalcTime = false;
    private long mLastLocationUpdate;

    public LocationService() {
    }

    public static void start(Context c) {
        Intent i = new Intent(c, LocationService.class);
        c.startService(i);
    }

    public static void triggerUpdate(Context c) {
        Intent i = new Intent(c, LocationService.class);
        i.setAction(UPDATE_LOCATION);
        c.startService(i);
    }


    @Override
    public IBinder onBind(Intent intent) {
        throw new UnsupportedOperationException("Not yet implemented");
    }

    @Override
    public void onCreate() {
        super.onCreate();
        if (ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_FINE_LOCATION) != PackageManager.PERMISSION_GRANTED
                && ActivityCompat.checkSelfPermission(this, Manifest.permission.ACCESS_COARSE_LOCATION) != PackageManager.PERMISSION_GRANTED) {
            stopSelf();
            return;
        }


        List<Times> times = Times.getTimes();
        for (Times t : times) {
            if (t.isAutoLocation()) {
                if (t instanceof CalcTimes)
                    mHasCalcTime = true;
                else if (t instanceof WebTimes)
                    mHasWebTime = true;

                if (mHasCalcTime && mHasWebTime) break;
            }
        }

        if (mHasCalcTime || mHasWebTime)
            initGPS();
        else
            this.stopSelf();
    }

    @Override
    public int onStartCommand(Intent intent, int flags, int startId) {
        if (!mHasCalcTime && !mHasWebTime) {
            stopSelf();
            return START_NOT_STICKY;
        }
        if (System.currentTimeMillis() - mLastLocationUpdate > 60 * 60 * 1000 || UPDATE_LOCATION.equals(intent.getAction())) {
            mLocationManager.requestSingleUpdate(LocationManager.NETWORK_PROVIDER, this, Looper.myLooper());
        }
        return START_STICKY;
    }

    @Override
    public void onDestroy() {
        mLocationManager.removeUpdates(this);
        super.onDestroy();
    }

    private void initGPS() {
        mLocationManager = (LocationManager) getSystemService(Context.LOCATION_SERVICE);
        mLocationManager.requestLocationUpdates(LocationManager.PASSIVE_PROVIDER, 1000 * 60 * 30, 5000, this);
    }

    @Override
    public void onLocationChanged(Location location) {
        mLastLocationUpdate = System.currentTimeMillis();
        final double lat = location.getLatitude();
        final double lng = location.getLongitude();
        final double elv = location.getAltitude();

        Cities.get().search(lat, lng, new Cities.Callback<List<Entry>>() {
            @SuppressWarnings("ConstantConditions")
            @Override
            public void onResult(List<Entry> result) {
                List<Times> times = Times.getTimes();
                mHasCalcTime = false;
                mHasWebTime = false;
                boolean updated = false;
                for (Times t : times) {
                    if (t.isAutoLocation()) {
                        for (Entry e : result) {
                            if (t.getSource() == e.getSource()) {
                                if (e.getSource() == Source.Calc) {
                                    String oldtimes[] = {t.getTime(0), t.getTime(1), t.getTime(2),
                                            t.getTime(3), t.getTime(4), t.getTime(5), t.getTime(6)};
                                    ((CalcTimes) t).getPrayTimes().setCoordinates(lat, lng, elv);
                                    t.setName(e.getName());
                                    for (int i = 0; i < oldtimes.length; i++) {
                                        if (oldtimes[i].equals(t.getTime(i))) {
                                            updated = true;
                                            break;
                                        }
                                    }
                                    mHasCalcTime = true;
                                } else {
                                    t.setName(e.getName());
                                    if (!((WebTimes) t).getId().equals(e.getKey())) {
                                        ((WebTimes) t).setId(e.getKey());
                                        updated = true;
                                        ((WebTimes) t).syncAsync();
                                    }
                                    mHasWebTime = true;
                                }
                            }
                        }
                    }
                }
                if (updated) {
                    Times.notifyDataSetChanged();
                    Times.setAlarms();
                    WidgetService.start(getApplication());
                }
            }
        });
    }


    @Override
    public void onStatusChanged(String s, int i, Bundle bundle) {

    }

    @Override
    public void onProviderEnabled(String s) {

    }

    @Override
    public void onProviderDisabled(String s) {

    }
}

File: features/times/src/main/java/com/metinkale/prayer/times/LocationService.kt
/*
 * Copyright (c) 2013-2023 Metin Kale
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.metinkale.prayer.times

import android.Manifest
import android.annotation.SuppressLint
import android.app.NotificationManager
import android.content.Context
import android.content.Intent
import android.content.pm.PackageManager
import android.location.Location
import android.location.LocationListener
import android.location.LocationManager
import android.os.Build
import androidx.core.app.ActivityCompat
import androidx.lifecycle.LifecycleService
import androidx.lifecycle.Observer
import androidx.lifecycle.asLiveData
import com.metinkale.prayer.App
import com.metinkale.prayer.receiver.AppEventManager
import com.metinkale.prayer.receiver.OnScreenOffListener
import com.metinkale.prayer.receiver.OnScreenOnListener
import com.metinkale.prayer.receiver.OnStartListener
import com.metinkale.prayer.times.times.DayTimesWebProvider
import com.metinkale.prayer.times.times.Times
import com.metinkale.prayer.times.times.setAlarms
import com.metinkale.prayer.times.utils.NotificationUtils
import dev.metinkale.prayertimes.calc.PrayTimes
import dev.metinkale.prayertimes.core.Entry
import dev.metinkale.prayertimes.core.sources.Source
import kotlinx.coroutines.flow.distinctUntilChanged
import kotlinx.coroutines.flow.filter
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.take

@SuppressLint("MissingPermission")
class LocationService : LifecycleService(), Observer<List<Entry>>, LocationListener,
    OnScreenOffListener, OnScreenOnListener {


    private val searchApi = OpenPrayerTimesSearchEndpoint()
    private lateinit var locationManager: LocationManager
    private lateinit var location: Location
    private var lastSearch: Long = 0
    override fun onStartCommand(intent: Intent?, flags: Int, startId: Int): Int {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            startForeground(
                NotificationUtils.getDummyNotificationId(),
                NotificationUtils.createDummyNotification(this)
            )
        }
        return super.onStartCommand(intent, flags, startId)
    }


    override fun onLocationChanged(location: Location) {
        val time = System.currentTimeMillis()
        if (!this::location.isInitialized || (location.distanceTo(this.location) > MIN_LOCATION_DISTANCE && time - lastSearch > MIN_LOCATION_AGE)) {
            this.location = location
            lastSearch = time
            val lat = location.latitude
            val lng = location.longitude
            searchApi.search(lat, lng)
        }
    }


    override fun onChanged(result: List<Entry>) {
        val lat = location.latitude
        val lng = location.longitude
        val elv = location.altitude

        for (t in Times.current) {
            if (t.autoLocation) {
                for (e in result) {
                    if (t.source === e.source) {
                        if (e.source === Source.Calc) {
                            Times.getTimesById(t.id).update {
                                it?.copy(name = e.localizedName(), key = it.key?.let {
                                    PrayTimes.deserialize(it)
                                        .copy(latitude = lat, longitude = lng, elevation = elv)
                                        .serialize()
                                })
                            }
                        } else {
                            Times.getTimesById(t.id).update {
                                t.copy(
                                    name = e.localizedName(),
                                    key = e.id,
                                    lat = e.lat ?: 0.0,
                                    lng = e.lng ?: 0.0
                                ).also { (it.dayTimes as? DayTimesWebProvider)?.syncAsync() }
                            }
                        }
                    }
                }
            }
        }

        if (Times.current.any { it.autoLocation }) {
            Times.setAlarms()
            AppEventManager.sendTimeTick()
        }
    }

    override fun onCreate() {
        super.onCreate()
        AppEventManager.register(this)
        Times.filter { it.none { it.autoLocation } }.take(1).asLiveData()
            .observe({ lifecycle }) { stopSelf() }

        locationManager = getSystemService(Context.LOCATION_SERVICE) as LocationManager


        onScreenOn()

        Times.map { it.any { it.autoLocation } }.distinctUntilChanged().filter { !it }
            .asLiveData().observe({ lifecycle }) {
                stopSelf()
            }

        searchApi.observe({ lifecycle }, this)
    }

    override fun onDestroy() {
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
            stopForeground(STOP_FOREGROUND_REMOVE)
            val notMan =
                getSystemService(Context.NOTIFICATION_SERVICE) as NotificationManager
            notMan.cancel(NotificationUtils.getDummyNotificationId())
        }
        AppEventManager.unregister(this)
        onScreenOff()
        super.onDestroy()
    }


    override fun onScreenOff() {
        locationManager.removeUpdates(this)
    }


    override fun onScreenOn() {
        locationManager.requestLocationUpdates(
            LocationManager.PASSIVE_PROVIDER,
            MIN_LOCATION_AGE,
            MIN_LOCATION_DISTANCE.toFloat(),
            this
        )
        locationManager.requestLocationUpdates(
            LocationManager.NETWORK_PROVIDER,
            MIN_LOCATION_AGE,
            MIN_LOCATION_DISTANCE.toFloat(),
            this
        )
    }


    companion object : OnStartListener {
        private const val MIN_LOCATION_AGE = 1000L * 60L * 30
        private const val MIN_LOCATION_DISTANCE = 5000

        private fun hasPermission(c: Context = App.get()): Boolean {
            return ActivityCompat.checkSelfPermission(
                c, Manifest.permission.ACCESS_FINE_LOCATION
            ) == PackageManager.PERMISSION_GRANTED || ActivityCompat.checkSelfPermission(
                c, Manifest.permission.ACCESS_COARSE_LOCATION
            ) == PackageManager.PERMISSION_GRANTED
        }

        private fun useAutoLocation(): Boolean = Times.current.any { it.autoLocation }

        override fun onStart() {
            Times.map { it.any { it.autoLocation } }.distinctUntilChanged().filter { it }
                .asLiveData()
                .observeForever { start() }
        }

        fun start() {
            if (!hasPermission() || !useAutoLocation()) return
            if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O) {
                App.get().startForegroundService(Intent(App.get(), LocationService::class.java))
            } else {
                App.get().startService(Intent(App.get(), LocationService::class.java))
            }
        }
    }


}